app/models/mhv_account.rb

```diff
@@ -0,0 +1,148 @@
+      transitions from: ALL_STATES, to: :unknown
+    end
+
+    event :check_terms_acceptance do
+      transitions from: ALL_STATES - [:ineligible], to: :needs_terms_acceptance, unless: :terms_and_conditions_accepted?
+    end
+
+    event :register do
+      transitions from: [:unknown, :register_failed], to: :registered
+    end
+
+    event :upgrade do
+      transitions from: [:unknown, :registered, :upgrade_failed], to: :upgraded
+    end
+  end
+
+  def create_and_upgrade!
+    create_mhv_account! unless preexisting_account?
+    upgrade_mhv_account!
+  end
+
+  def terms_and_conditions_accepted?
+    terms_and_conditions_accepted.present?
+  end
+
+  def preexisting_account?
+    user&.mhv_correlation_id.present?
+  end
+
+  private
+
+  def terms_and_conditions_accepted
+    @terms_and_conditions_accepted ||=
+      TermsAndConditionsAcceptance.joins(:terms_and_conditions)
+                                  .includes(:terms_and_conditions)
+                                  .where(terms_and_conditions: { latest: true, name: 'mhv_account_terms' })
+                                  .where(user_uuid: user.uuid).limit(1).first
+  end
+
+  def params_for_registration
+    {
+      icn: user.icn,
+      is_patient: va_patient?,
+      is_veteran: veteran?,
+      address1: user.va_profile.address.street,
+      city: user.va_profile.address.city,
+      state: user.va_profile.address.state,
+      zip: user.va_profile.address.postal_code,
+      country: user.va_profile.address.country,
+      province: nil, # TODO: We need to determine if this is something that could actually happen (non USA)
+      email: user.email,
+      home_phone: user.va_profile.home_phone,
+      sign_in_partners: 'VETS.GOV',
+      terms_version: terms_and_conditions_accepted.terms_and_conditions.version,
+      terms_accepted_date: terms_and_conditions_accepted.created_at
+    }
+  end
+
+  def params_for_upgrade
+    {
+      user_id: user.mhv_correlation_id || @new_mhv_correlation_id,
+      form_signed_date_time: terms_and_conditions_accepted.created_at,
+      terms_version: terms_and_conditions_accepted.terms_and_conditions.version
+    }
+  end
+
+  def user
+    @user ||= User.find(user_uuid)
+  end
+
+  def mhv_account_eligible?
+    va_patient?
+  end
+
+  def va_patient?
+    # TODO: This needs to be changed to check if ICN is within a certain range.
+    user&.icn.present?
+  end
+
+  def veteran?
+    # TODO: this field is derived from eMIS and might have pending ATO considerations for us to use it.
+    true
+  end
+
+  def create_mhv_account!
+    if may_register?
+      client_response = mhv_ac_client.post_register(params_for_registration)
+      if client_response[:api_completion_status] == 'Successful'
+        @new_mhv_correlation_id = client_response[:correlation_id]
+        # TODO: figure out how to bust and refresh the MVI cache
+        self.registered_at = Time.current
+        register!
```

Comment:
I'd be inclined to call this `registered!` since you just did the registration above.